lib/spack/spack/environment/environment.py

```diff
@@ -556,9 +556,10 @@ def _is_dev_spec_and_has_changed(spec):
         # Not installed -> nothing to compare against
         return False
 
-    _, record = spack.store.STORE.db.query_by_spec_hash(spec.dag_hash())
-    mtime = fs.last_modification_time_recursive(dev_path_var.value)
-    return mtime > record.installation_time
+    # hook so packages can use to write their own method for checking the dev_path
+    # use package so attributes about concretization such as variant state can be
+    # utilized
```

Comment:
```suggestion
    # check whether the package source was changed since the last install
    # default is to compare all modification times against installation time
    # can be overridden on a per-package basis.
```
Doesn't need to be exactly that, but I think those are the three important pieces of information to put here. As a package method, it's obvious that it has access to the full package state.

Reply:
At the time I was distinguishing/justifying between package_class and package to myself. As you say, obvious now.